substrate-archive/src/types.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of substrate-archive.
+
+// substrate-archive is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// substrate-archive is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+use sp_runtime::traits::Block as BlockT;
+use substrate_archive_common::error::Result;
+
+pub trait ThreadPool: Send + Sync {
+    type In: Send + Sync + std::fmt::Debug;
+    type Out: Send + Sync + std::fmt::Debug;
+    /// Adds a task to the threadpool.
+    /// Should not block!
+    fn add_task(&self, d: Vec<Self::In>, tx: flume::Sender<Self::Out>) -> Result<usize>;
+}
+
+#[async_trait::async_trait(?Send)]
+pub trait Archive<B: BlockT + Unpin>
+where
```

Comment:
Would make sense to remove the `generic trait param`  on the entire trait and just add the bounds on context because it's not needed by the other methods:

```rust
trait Archive {
    // other methods...

    fn context<B>(&self) -> Result<super::actors::ActorContext<B>>
    where 
           B: BlockT + Unpin,
           <B as BlockT>::Hash: Unpin;
```

FWIW, it would likely require some additional refactoring and probably not worth the effort